drivers/input/sbutton.c

```diff
@@ -0,0 +1,269 @@
+
+  if (state == KEY_PRESS && !priv->pressed)
+    {
+      iinfo("Button pressed\n");
+
+      priv->pressed = true;
+      priv->start   = clock_systime_ticks();
+    }
+  else
+    {
+      if (state == KEY_RELEASE && priv->pressed)
+        {
+          uint32_t elapsed;
+
+          iinfo("Button released\n");
+
+          priv->pressed = false;
+          priv->end = clock_systime_ticks();
+
+          elapsed = priv->end - priv->start;
+          if (elapsed < MSEC2TICK(KEY_THRESH_MS))
+            {
+              iinfo("KEY_1\n");
+              keyboard_event(&priv->lower, ASCII_A, KEYBOARD_PRESS);
+            }
+          else
+            {
+              iinfo("KEY_2\n");
+              keyboard_event(&priv->lower, ASCII_B, KEYBOARD_PRESS);
+            }
+        }
+    }
+
+  /* Unlock and return */
+
+  nxmutex_unlock(&priv->lock);
+}
+
+/****************************************************************************
+ * Name: sbutton_interrupt
+ ****************************************************************************/
+
+static int sbutton_interrupt(int irq, FAR void *context, FAR void *arg)
+{
+  FAR struct sbutton_dev_s *priv = (FAR struct sbutton_dev_s *)arg;
+  int                        ret;
+
+  /* Let the event worker know that it has an interrupt event to handle
+   * It is possible that we will already have work scheduled from a
+   * previous interrupt event.  That is OK we will service all the events
+   * in the same work job.
+   */
+
+  if (work_available(&priv->work))
+    {
+      ret = work_queue(HPWORK, &priv->work, sbutton_worker, priv, 0);
+      if (ret != 0)
+        {
+          ierr("ERROR: Failed to queue work: %d\n", ret);
+        }
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: sbutton_register
+ *
+ * Description:
+ *   Configure the Single Button Dual Key Keyboard to use the provided
+ *   instance.  This will register the driver as /dev/kbdN where N is the
+ *   minor device number, as well as a joystick at joydevname
+ *
+ * Input Parameters:
+ *   config      - Persistent board configuration data
+ *   kbdminor    - The keyboard input device minor number
+ *
+ * Returned Value:
+ *   Zero is returned on success.  Otherwise, a negated errno value is
+ *   returned to indicate the nature of the failure.
+ *
+ ****************************************************************************/
+
+int sbutton_register(FAR const struct sbutton_config_s *config,
+                     char kbdminor)
+{
+  FAR struct sbutton_dev_s *priv;
+  char                      kbddevname[DEV_NAMELEN];
+  int                       ret;
+
+  /* Debug Sanity Checks */
+
+  DEBUGASSERT(config != NULL);
+  DEBUGASSERT(config->attach != NULL && config->enable != NULL &&
```

Comment:
I think checking each of those in separate `DEBUGASSERT` would be more readable. But that's just me.

Reply:
done